packages/tailwindcss/src/index.ts

```diff
@@ -188,9 +191,307 @@ async function parseCss(
         )
       }
 
-      customUtilities.push((designSystem) => {
-        designSystem.utilities.static(name, () => structuredClone(node.nodes))
-      })
+      // Functional utilities. E.g.: `tab-size-*`
+      if (IS_VALID_FUNCTIONAL_UTILITY_NAME.test(name)) {
+        customUtilities.push((designSystem) => {
+          designSystem.utilities.functional(name.slice(0, -2), (candidate) => {
+            let ast = structuredClone(node.nodes)
+
+            // A value is required for functional utilities, if you want to
+            // accept just `tab-size`, you'd have to use a static utility.
+            if (candidate.value === null) return
+
+            // Whether `value(…)` was used
+            let usedValueFn = false
+
+            // Whether any of the declarations successfully resolved a `value(…)`.
+            // E.g:
+            // ```css
+            // @utility tab-size-* {
+            //   tab-size: value(integer);
+            //   tab-size: value(--tab-size);
+            //   tab-size: value([integer]);
+            // }
+            // ```
+            // Any of these `tab-size` declarations have to resolve to a valid
+            // in order to make the utility valid.
+            let resolvedValueFn = false
+
+            // The resolved value type, e.g.: `integer`
+            let resolvedValueType = null as string | null
+
+            // Whether `modifier(…)` was used
+            let usedModifierFn = false
+
+            // Whether any of the declarations successfully resolved a `modifier(…)`
+            let resolvedModifierFn = false
+
+            walk(ast, (node, { replaceWith: replaceDeclarationWith }) => {
+              if (node.kind !== 'declaration') return
+              if (!node.value) return
+
+              let valueAst = ValueParser.parse(node.value.replace(/\s+\*/g, '*'))
+              let result =
+                ValueParser.walk(valueAst, (valueNode, { replaceWith }) => {
+                  if (valueNode.kind !== 'function') return
+
+                  // Value function, e.g.: `value(integer)`
+                  if (valueNode.value === 'value') {
+                    usedValueFn = true
+
+                    for (let arg of valueNode.nodes) {
+                      // Resolving theme value, e.g.: `value(--color)`
+                      if (
+                        candidate.value?.kind === 'named' &&
```

Comment:
We know that `candidate.value` is not `null` because of the check on top. Maybe we can store `value` in a separate value at that point so we can avoid the repeated property access and nullability checks in here?